sdks/python/apache_beam/transforms/external.py

```diff
@@ -42,6 +44,63 @@
   grpc = None
 # pylint: enable=wrong-import-order, wrong-import-position, ungrouped-imports
 
+DEFAULT_EXPANSION_SERVICE = 'localhost:8097'
+FILTERED_CODERS = {
+  'beam:coder:length_prefix:v1',
+}
+
+
+def iter_urns(coder, context=None):
+  yield coder.to_runner_api_parameter(context)[0]
+  for child in coder._get_component_coders():
+    for urn in iter_urns(child, context):
+      yield urn
+
+
+class External(ptransform.PTransform):
+  """
+  Base class for transforms that produce external transforms.
+  """
+  _urn = None
+
+  def __init__(self, expansion_service=None):
+    self.expansion_service = expansion_service or DEFAULT_EXPANSION_SERVICE
+
+  def get_config_args(self):
+    """
+    :return: dictionary of encoded configuration values to pass to
+             ExternalConfigurationPayload
+    """
+    raise NotImplementedError
+
+  @classmethod
+  def config_value(cls, obj, coder):
+    """
+    Helper to create a ConfigValue with an encoded value.
+    """
+    return ConfigValue(
+      coder_urn=[urn for urn in iter_urns(coder)
+                 if urn not in FILTERED_CODERS],
+      payload=coder.encode(obj))
+
+  def expand(self, pbegin):
+    args = self.get_config_args()
+
+    payload = ExternalConfigurationPayload(configuration=args)
+    return pbegin.apply(
+        ExternalTransform(
+            self._urn,
+            payload.SerializeToString(),
+            self.expansion_service))
+
+
+class ExternalRead(ptransform.PTransform):
+  def expand(self, pbegin):
+    if not isinstance(pbegin, pvalue.PBegin):
```

Comment:
I'm not sure if this check is worth having a separate class, especially as we move to more modular IO that does not necessarily have to be at the root.

Reply:
I agree.  I left it there to match the previous behavior, but I'd love to drop down to one class.

Reply:
Would also favor removing this class. It shouldn't matter if this is a read or not. Perhaps we could encode this in the `External` class whether the transform requires an Impulse or not?

Reply:
Removed `ExternalRead`.